File: apps/cli/src/commands/copilot.ts
/**
 * ortho guardrails|decide|plan|refactor — Engineering copilot commands.
 * Bridges to copilot.py (real CliCommands: ArchitectureEnforcer,
 * DecisionEngine, FeaturePlanner, RefactoringAdvisor).
 */

import { Command } from "commander";
import { runPython } from "./pybridge";

const COPILOT_CLI = "apps/cli/src/commands/copilot.py";

/**
 * Run the bridge and map failures. A failed *report* (nonexistent path,
 * rejected intent) has already been printed to stdout by the Python side
 * before it exits 1 — re-printing the generic "exit code 1" rejection
 * would mislabel an expected failure report as a CLI crash, so only
 * spawn-level errors (missing python, etc.) get an "Error:" line.
 */
async function runCopilot(args: string[]): Promise<void> {
  try {
    await runPython(COPILOT_CLI, args);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    if (!/exit code \d+/.test(message)) {
      console.error("Error:", message);
    }
    process.exit(1);
  }
}

function requireIntent(intent: string, command: string): string {
  if (!intent || intent.trim().length === 0) {
    console.error(`Error: ${command} requires a non-empty intent.`);
    process.exit(1);
  }
  return intent;
}

export const reviewCommand = new Command()
  .command("review [path]")
  .description("Unified review: architecture violations + decision summary, one scan")
  .option("--severity <level>", "error or warning (default: all)")
  .action(async (path?: string, options?: { severity?: string }) => {
    if (options?.severity && !["error", "warning"].includes(options.severity)) {
      console.error(`Error: --severity must be 'error' or 'warning', got '${options.severity}'`);
      process.exit(1);
    }
    const args = ["review", "--path", path || process.cwd()];
    if (options?.severity) args.push("--severity", options.severity);
    await runCopilot(args);
  });

export const guardrailsCommand = new Command()
  .command("guardrails [path]")
  .description("Check architecture violations (layer boundaries, cycles, module size)")
  .option("--severity <level>", "error or warning (default: all)")
  .option("--against-branch <name>", "Restrict violations to files changed vs. this local branch")
  .action(async (path?: string, options?: { severity?: string; againstBranch?: string }) => {
    if (options?.severity && !["error", "warning"].includes(options.severity)) {
      console.error(`Error: --severity must be 'error' or 'warning', got '${options.severity}'`);
      process.exit(1);
    }
    const args = ["guardrails", "--path", path || process.cwd()];
    if (options?.severity) args.push("--severity", options.severity);
    if (options?.againstBranch) args.push("--against-branch", options.againstBranch);
    await runCopilot(args);
  });

export const decideCommand = new Command()
  .command("decide <intent>")
  .description("Decision support: change-impact + guardrail aggregation (intent = text or file path)")
  .option("--scan-path <dir>", "Directory to scan (default: current directory)")
  .option("--confidence <threshold>", "Minimum confidence 0.0–1.0")
  .action(async (intent: string, options?: { scanPath?: string; confidence?: string }) => {
    requireIntent(intent, "decide");
    if (options?.confidence) {
      const conf = parseFloat(options.confidence);
      if (isNaN(conf) || conf < 0.0 || conf > 1.0) {
        console.error(`Error: --confidence must be 0.0–1.0, got '${options.confidence}'`);
        process.exit(1);
      }
    }
    const args = ["decide", intent, "--scan-path", options?.scanPath || process.cwd()];
    if (options?.confidence) args.push("--confidence", options.confidence);
    await runCopilot(args);
  });

export const planCommand = new Command()
  .command("plan <intent>")
  .description("Feature planning: classified intent -> implementation paths")
  .option("--scan-path <dir>", "Directory to scan (default: current directory)")
  .action(async (intent: string, options: { scanPath?: string }) => {
    requireIntent(intent, "plan");
    await runCopilot(["plan", intent, "--scan-path", options.scanPath ?? process.cwd()]);
  });

export const refactorCommand = new Command()
  .command("refactor [path]")
  .description("Refactoring findings: bloat, tight coupling, circular dependencies")
  .action(async (path?: string) => {
    await runCopilot(["refactor", "--path", path ?? process.cwd()]);
  });

export const feedbackCommand = new Command()
  .command("feedback <decision> <findingKey>")
  .description(
    'Record accept/reject on a finding ("{rule_id} {location}" from guardrails/decide/review output) -- future runs cite the decision and reason instead of just "seen before"'
  )
  .option("--path <dir>", "Repository path (default: current directory)")
  .option("--reason <text>", "Why (shown on future runs if rejected)")
  .action(async (decision: string, findingKey: string, options?: { path?: string; reason?: string }) => {
    if (decision !== "accept" && decision !== "reject") {
      console.error(`Error: decision must be 'accept' or 'reject', got '${decision}'`);
      process.exit(1);
    }
    requireIntent(findingKey, "feedback");
    const args = ["feedback", decision, findingKey, "--path", options?.path || process.cwd()];
    if (options?.reason) args.push("--reason", options.reason);
    await runCopilot(args);
  });

export const askCommand = new Command()
  .command("ask <question>")
  .description(
    "Repository Understanding: structural Q&A grounded in the real call/import graph (e.g. \"how does auth work\") -- not vector search, answers only from real graph evidence"
  )
  .option("--scan-path <dir>", "Directory to scan (default: current directory)")
  .action(async (question: string, options?: { scanPath?: string }) => {
    requireIntent(question, "ask");
    await runCopilot(["ask", question, "--scan-path", options?.scanPath || process.cwd()]);
  });

export const orchestrateCommand = new Command()
  .command("orchestrate <intent>")
  .description(
    "Chain plan+decide+review into one composed report for an intent. Does not write code, approve, or merge -- that's the developer's / LLM's job; Ortho advises."
  )
  .option("--scan-path <dir>", "Directory to scan (default: current directory)")
  .action(async (intent: string, options?: { scanPath?: string }) => {
    requireIntent(intent, "orchestrate");
    await runCopilot(["orchestrate", intent, "--scan-path", options?.scanPath || process.cwd()]);
  });

export const crossRepoCommand = new Command()
  .command("cross-repo <paths...>")
  .description(
    "Shared/reusable code across 2-5 real repos, via real AST-structural similarity -- not naming-based guesswork"
  )
  .option("--threshold <value>", "Similarity threshold 0.0-1.0 (default: 0.7)")
  .action(async (paths: string[], options?: { threshold?: string }) => {
    if (paths.length < 2) {
      console.error(`Error: cross-repo needs at least 2 repo paths, got ${paths.length}`);
      process.exit(1);
    }
    if (options?.threshold) {
      const t = parseFloat(options.threshold);
      if (isNaN(t) || t < 0.0 || t > 1.0) {
        console.error(`Error: --threshold must be 0.0–1.0, got '${options.threshold}'`);
        process.exit(1);
      }
    }
    const args = ["cross-repo", ...paths];
    if (options?.threshold) args.push("--threshold", options.threshold);
    await runCopilot(args);
  });

export const memoryCommand = new Command()
  .command("memory <query>")
  .description("Search workflow_run artifacts in ContextHub (learned from past guardrails/decide/plan/refactor runs)")
  .option("--repo-path <dir>", "Repository path (default: current directory)")
  .action(async (query: string, options?: { repoPath?: string }) => {
    requireIntent(query, "memory");
    const args = [query, "--repo-path", options?.repoPath || process.cwd()];
    await runPython("apps/cli/src/commands/memory_search.py", args);
  });

File: apps/cli/src/commands/copilot.py
"""CLI bridge for the four engineering-copilot commands.

ortho guardrails/decide/plan/refactor -> CliCommands (packages/cli-commands).
Mirrors context.py's bridge pattern: sys.path bootstrap, argparse _main(),
spawned by copilot.ts via pybridge.ts.

--path/--scan-path are required at this level: the TS side always passes
an explicit directory (defaulting to the user's cwd), keeping the
no-unbounded-scan discipline from tasks 017/019 mechanical.
"""

import sys
from pathlib import Path

# Spawned as a bare script by the TS CLI — add dependencies to path FIRST
_PROJECT_ROOT = Path(__file__).resolve().parents[4]
for _pkg in (
    _PROJECT_ROOT / "shared" / "storage" / "src",
    _PROJECT_ROOT / "packages" / "repo-intelligence" / "src",
    _PROJECT_ROOT / "packages" / "arch-intelligence" / "src",
    _PROJECT_ROOT / "packages" / "impact-analysis" / "src",
    _PROJECT_ROOT / "packages" / "change-planner" / "src",
    _PROJECT_ROOT / "packages" / "feature-planner" / "src",
    _PROJECT_ROOT / "packages" / "refactoring-advisor" / "src",
    _PROJECT_ROOT / "packages" / "arch-guardrails" / "src",
    _PROJECT_ROOT / "packages" / "decision-engine" / "src",
    _PROJECT_ROOT / "packages" / "context-hub" / "src",
    _PROJECT_ROOT / "packages" / "cli-commands" / "src",
):
    if str(_pkg) not in sys.path:
        sys.path.insert(0, str(_pkg))


def _main() -> None:
    import argparse

    parser = argparse.ArgumentParser(description="ortho engineering copilot")
    subparsers = parser.add_subparsers(dest="action", required=True)

    review_parser = subparsers.add_parser("review", help="Unified guardrails + decision summary")
    review_parser.add_argument("--path", required=True, help="Directory to scan")
    review_parser.add_argument("--severity", choices=["error", "warning"], help="Filter violations by severity")

    guardrails_parser = subparsers.add_parser("guardrails", help="Architecture violation check")
    guardrails_parser.add_argument("--path", required=True, help="Directory to scan")
    guardrails_parser.add_argument("--severity", choices=["error", "warning"], help="Filter violations by severity")

    decide_parser = subparsers.add_parser("decide", help="Decision support")
    decide_parser.add_argument("intent", help="Free-text intent or a file path")
    decide_parser.add_argument("--scan-path", required=True, help="Directory to scan")
    decide_parser.add_argument("--confidence", help="Minimum confidence threshold (0.0-1.0)")

    plan_parser = subparsers.add_parser("plan", help="Feature implementation paths")
    plan_parser.add_argument("intent", help="Free-text feature intent")
    plan_parser.add_argument("--scan-path", required=True, help="Directory to scan")

    refactor_parser = subparsers.add_parser("refactor", help="Refactoring findings")
    refactor_parser.add_argument("--path", required=True, help="Directory to scan")

    feedback_parser = subparsers.add_parser("feedback", help="Record accept/reject on a finding")
    feedback_parser.add_argument("decision", choices=["accept", "reject"], help="accept or reject")
    feedback_parser.add_argument("finding_key", help='"{rule_id} {location}" as shown in guardrails/decide/review output')
    feedback_parser.add_argument("--path", required=True, help="Repository path")
    feedback_parser.add_argument("--reason", default="", help="Why (shown on future runs if rejected)")

    ask_parser = subparsers.add_parser("ask", help="Repository Understanding: structural Q&A")
    ask_parser.add_argument("question", help="Question, e.g. 'how does auth work'")
    ask_parser.add_argument("--scan-path", required=True, help="Directory to scan")

    orchestrate_parser = subparsers.add_parser("orchestrate", help="Chain plan+decide+review into one report")
    orchestrate_parser.add_argument("intent", help="Free-text intent")
    orchestrate_parser.add_argument("--scan-path", required=True, help="Directory to scan")

    cross_repo_parser = subparsers.add_parser("cross-repo", help="Shared/reusable code across 2-5 real repos")
    cross_repo_parser.add_argument("paths", nargs="+", help="2-5 repository paths to compare")
    cross_repo_parser.add_argument("--threshold", type=float, default=0.7, help="Similarity threshold 0.0-1.0")

    args = parser.parse_args()

    from cli_commands.commands import CliCommands

    commands = CliCommands()
    if args.action == "review":
        kwargs = {}
        if hasattr(args, "severity") and args.severity:
            kwargs["severity_filter"] = args.severity
        report = commands.review(args.path, **kwargs)
    elif args.action == "guardrails":
        kwargs = {}
        if hasattr(args, "severity") and args.severity:
            kwargs["severity_filter"] = args.severity
        report = commands.guardrails(args.path, **kwargs)
    elif args.action == "decide":
        kwargs = {"scan_path": args.scan_path}
        if hasattr(args, "confidence") and args.confidence:
            try:
                conf = float(args.confidence)
            except ValueError:
                parser.error(f"--confidence must be a float, got '{args.confidence}'")
            if not (0.0 <= conf <= 1.0):
                parser.error(f"--confidence must be 0.0–1.0, got {conf}")
            kwargs["confidence_threshold"] = conf
        report = commands.decide(args.intent, **kwargs)
    elif args.action == "plan":
        report = commands.plan(args.intent, scan_path=args.scan_path)
    elif args.action == "feedback":
        report = commands.feedback(args.path, args.finding_key, args.decision, args.reason)
    elif args.action == "ask":
        report = commands.ask(args.scan_path, args.question, scan_path=args.scan_path)
    elif args.action == "orchestrate":
        report = commands.orchestrate(args.intent, scan_path=args.scan_path)
    elif args.action == "cross-repo":
        report = commands.cross_repo(args.paths, threshold=args.threshold)
    else:
        report = commands.refactor(args.path)

    # Windows console encoding: use binary mode to bypass cp1252 codec
    # and write UTF-8 bytes directly to stdout (which is inherited from the
    # TS layer and goes to the real terminal/pipe).
    output = report.title + "\n\n" + report.content + "\n"
    sys.stdout.buffer.write(output.encode("utf-8"))
    sys.stdout.buffer.flush()
    sys.exit(0 if report.success else 1)


if __name__ == "__main__":
    _main()
